lang_java/tree_sitter/grammar_to_tree_sitter_ast.py

```diff
@@ -0,0 +1,128 @@
+import json
+
+with open("java_grammar.json", 'r') as fs:
+    data = json.load(fs)
+
+codegen = ""
+
+for a, aval in data["rules"].items():
+    if aval["type"] == "CHOICE":
+        codegen += '''
+type {type} = '''.format(type=a)
+        for b in aval["members"]:
+            if b["type"] not in ["PREC_DYNAMIC", "PREC_LEFT", "SEQ", "STRING", "ALIAS"]:
+                codegen +='''
+  | {constructor} of {field}'''.format(constructor=b['name'].split("_")[0].capitalize(),field=b['name'])
+        codegen += "\n"
+    elif aval["type"] == "TOKEN": #for literals
+        pass
+    elif aval["type"] == "REPEAT": #only for program.modifiers
+        pass
+    elif aval["type"] == "STRING": #don't need
+        pass
+    elif aval["type"] == "PREC":
+        pass
+    elif aval["type"] == "PREC_RIGHT":
+        pass
+    elif aval["type"] == "PREC_LEFT":
+        pass
+    elif aval["type"] == "SEQ":
+        codegen +='''
+and {type} = {{'''.format(type=a)    
+        for b in aval["members"]:
+            if b["type"] == "CHOICE":
+                for c in b['members']:
+                    if c['type'] not in ["SYMBOL","BLANK","STRING"]:
+                        if c['type'] == 'CHOICE':
+                            for d in c['members']: 
+                                for e, e_val in d.items():
+                                    if e == "members":
+                                        for f in e_val:
+                                            if f['type'] == 'SYMBOL':
+                                                codegen += "{param} * ".format(param=f['name'])
+                                            else:
+                                                for g, gval in f['content'].items():
+                                                    if g == "members":
+                                                        for h in gval:
+                                                            if h['type'] == 'SYMBOL':
+                                                                codegen += "{param} repeat;\n".format(param=h['name'])
+                        elif c['type'] == 'SEQ':
```

Comment:
I feel this should be a recursive function otherwise it unreadable. I think you are unrolling a recursive function unnecessarily